content/posts/2022-05-13-calculating-type-sets.markdown

```diff
@@ -0,0 +1,498 @@
+
+Surprisingly, there is also a simple form of negation. It happens implicitly,
+because a type can not have two methods with the same name, but different
+types. So, implicitly, if a type has a method `X()`, it does *not* have a
+method `X() int`, for example[^5]:
+
+```go
+type X interface { X() }
+type NotX interface{ X() int }
+```
+
+This meant that you could express interfaces which could never be satisfied by
+any type (i.e. which described an empty type set):
+
+```go
+interface{ X; NotX }
+```
+
+[The compiler rejects such interfaces](https://go.dev/play/p/r4kpXNynscX).
+
+So, we pretty much had a language to build boolean formulas, where every
+formula had the form
+
+    X∧Y∧¬Z
+
+Checking if a type implements an interface meant to check if it is in the set
+described by that interfaces. Which meant checking if the corresponding formula
+is satisfied.
+
+And checking if an interface is invalid meant checking if the set it describes
+is empty. Which meant checking if the formula is *satisfiable*. This was easy,
+though. If you look closely, those formulas are in DNF and as we noted above,
+solving SAT on a formula given in DNF is easy.
+
+## Adding unions
+
+Go 1.18 extends the interface syntax. For our purposes, the important addition
+is the `|` operator:
+
+```go
+type S interface{
+    A | B
+}
+```
+
+This represents the set of all types which are in the *union* of the type sets
+`A` and `B` - that is, the set of all types which are in `A` *or* in `B` (or
+both).
+
+This means our language of expressible formulas now also includes a
+`∨`-operator. What's more, the form of our formula is now a *Conjunctive*
+normal form - every line is a term of `∨` and the lines are connected by `∧`:
+
+```go
+type X interface { X() }
+type NotX interface{ X() int }
+type Y interface { Y() }
+type NotY interface{ Y() int }
+type Z interface { Z() }
+type NotZ interface{ Z() int }
+
+// (¬X ∨ Z) ∧ (Y ∨ Z) ∧ (X ∨ ¬Y)
+type S interface {
+    NotX | Z
+    Y | Z
+    X | NotY
+}
+```
+
+There is one detail which becomes important here. There are types which
+*neither* implement `X` *nor* do they implement `NotX` - if they don't have an
+`X` method at all, for example. If we want to really prove that calculating
+these new type sets is NP complete, we need to fix that. We want that a type
+set is empty *exactly* if its formula is not satisfiable. Boolean variables are
+always either `true` or `false`, though.
+
+“Luckily”, our new operator gives us a way to fix that:
+
+```go
+type TertiumNonDatur interface {
+    X | NotX
+    Y | NotY
+    Z | NotZ
+}
+
+// (¬X ∨ Z) ∧ (Y ∨ Z) ∧ (X ∨ ¬Y)
+type S interface {
+    TertiumNonDatur
+
+    NotX | Z
+    Y | Z
+    X | NotY
+}
+```
+
+Now, any type we consider necessarily has either an `X()` or an `X() int`
+method, so either implements `X`, or implements `NotX` (and never both).
+
+So, checking if these general type sets are empty is NP complete.
```

Comment:
Make this more explicit?
```suggestion
So because checking that an interface has an empty type set is equivalent to checking that a boolean formula in CNF has no solutions, which is NP complete as described above.
```